### PR TITLE
Persist selectedTag filtering of the reading list 

### PR DESCRIPTION
**Pull request by tinkermakar**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This change to Persist selectedTag filtering of the reading list when coming back after reading an article. This is achieved via a query param. Since the project does not use react-router-dom, the selected tag is retrieved via 2 lines of vanilla JS.

## Related Tickets & Documents

Discussion: https://github.com/forem/forem/discussions/20585

## QA Instructions, Screenshots, Recordings

1. Go to the reading list
1. Filter by a tag -- the `selectedTag` query parameter will be added to the URL
1. Refresh or open an article and then press back -- in all scenarios the list of articles will persist the filtering by selected tag
1. Press All Tags -- `selectedTag` query parameter should disappear, all articles should be listed

image

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above._

- [ ] Yes
- [ ] No, and this is why
- [ X ] I need help with writing tests

### CODE REVIEW

**Comment by tinkermakar** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by tinkermakar** [comment]
@lightalloy @maestromac the cypres failures in the previous CI runs were not my fault, were they?

**Comment by maestromac** [comment]
@tinkermakar no that's unrelated. I will address it shortly

**Comment by tinkermakar** [comment]
@maestromac still not my fault, is it? Seems a Ruby issue, not preact

**Comment by tinkermakar** [comment]
@lightalloy @maestromac I need help or at least some guidance with how to pass diff unit tests -- the lines I have added are all related to website URL (to fetch a query param), which is not a thing in the existing test environment.

**Comment by lightalloy** [comment]
@tinkermakar I believe there are cypress tests for filtering reading list by tag in `cypress/e2e/seededFlows/readingListFlows/readingList.spec.js`  , you can use them as examples and add test(s) for your case there.

As I see, `Cypress.config().baseUrl` is used there. Not unit tests (you asked for) but I think adding an integration test will be enough in this case.

**Comment by tinkermakar** [comment]
@lightalloy e2e added, thank you.

I also had to fix a bug -- in mobile view you cannot filter the reading list by a tag and then filter back by all -- it tries to fetch 'all tags' as a distinct tag and, natuarlly, gets 0 items from the backend. The bug can be replicated on prod https://dev.to right now

**Comment by tinkermakar** [comment]
@lightalloy sorry for bothering -- seems like the initial issue with the tests is back again

**Comment by tinkermakar** [comment]
@maestromac can you do your magic again, please?

**Comment by tinkermakar** [comment]
@lightalloy sorry for bothering again -- seems like we're all good to merge, aren't we?

File: app/javascript/readingList/readingList.jsx [hunk 1]
```diff
@@ -69,12 +70,20 @@ export class ReadingList extends Component {
     this.clearSelectedTags = clearSelectedTags.bind(this);
   }
 
-  componentDidMount() {
+  async componentDidMount() {
```
Comment (maestromac) [inline app/javascript/readingList/readingList.jsx:73]: Can you explain why some of these method needs to be async?

**Review by maestromac — APPROVED** [review]
LGMT! Tested locally.

File: app/javascript/readingList/readingList.jsx (hunk 1 above)
Comment (tinkermakar) [reply app/javascript/readingList/readingList.jsx:73]: so as `performInitialSearch` and `selectTag` did not run in parallel -- the former must complete before the latter starts, or else it malfunctions. They both result in updating the list of posts in the view. I could theoretically make my change as a part of  `performInitialSearch`, but reusing the existing `selectTag` methods seems a less intrusive and DRY way to go.

**Comment by tinkermakar** [comment]
I see room for further corrections here -- I'll open another PR after I use the reading list more.

**Pull request merged**